### PR TITLE
add example mnist_net2net.py

### PR DESCRIPTION
**Pull request by dolaameng**
Example based on paper ["Net2Net: Accelerating Learning via Knowledge Transfer"](http://arxiv.org/abs/1511.05641). The results on MNIST showed the advantages of 

1. Net2WiderNet over Random-Padding
2. Net2DeeperNet over Random-Initialization

### CODE REVIEW

File: examples/mnist_net2net.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+from __future__ import print_function
+
+from keras.models import Sequential
+from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
+from keras.utils import np_utils
+from keras.datasets import mnist
+
+import numpy as np
+
+np.random.seed(1337)
```
Comment (fchollet) [inline examples/mnist_net2net.py]: For this to be effective it should happen before Keras imports

File: examples/mnist_net2net.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+input_shape = (1, 28, 28)  # image shape
+nb_class = 10  # number of class
+
+# load and pre-process data
+
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Unnecessary whitespace

File: examples/mnist_net2net.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+(train_x, train_y), (validation_x, validation_y) = mnist.load_data()
+
+
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Unnecessary whitespace

File: examples/mnist_net2net.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+def preprocess_input(x):
+    return x.reshape((-1, ) + input_shape) / 255.
+
+
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Unnecessary whitespace

File: examples/mnist_net2net.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+train_x, validation_x = map(preprocess_input, [train_x, validation_x])
+train_y, validation_y = map(preprocess_output, [train_y, validation_y])
+print("Loading MNIST data...")
+print("train_x shape:", train_x.shape, "train_y shape:", train_y.shape)
+print("validation_x shape:", validation_x.shape,
+      "validation_y shape", validation_y.shape)
+
+
+# knowledge transfer algorithms
+
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Unnecessary whitespace

File: examples/mnist_net2net.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    return np_utils.to_categorical(y)
+
+train_x, validation_x = map(preprocess_input, [train_x, validation_x])
+train_y, validation_y = map(preprocess_output, [train_y, validation_y])
+print("Loading MNIST data...")
+print("train_x shape:", train_x.shape, "train_y shape:", train_y.shape)
+print("validation_x shape:", validation_x.shape,
+      "validation_y shape", validation_y.shape)
+
+
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Unnecessary whitespace

File: examples/mnist_net2net.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+# knowledge transfer algorithms
+
+def wider2net_conv2d(teacher_w1, teacher_b1, teacher_w2, new_width, init):
+    """Get initial weights for a wider conv2d layer with a bigger nb_filter,
+    by 'random-padding' or 'net2wider'.
+
+    # Auguments
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Typo

File: examples/mnist_net2net.py [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+print("Loading MNIST data...")
+print("train_x shape:", train_x.shape, "train_y shape:", train_y.shape)
+print("validation_x shape:", validation_x.shape,
+      "validation_y shape", validation_y.shape)
+
+
+# knowledge transfer algorithms
+
+def wider2net_conv2d(teacher_w1, teacher_b1, teacher_w2, new_width, init):
+    """Get initial weights for a wider conv2d layer with a bigger nb_filter,
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Prefer using `'''` instead of `"""`

File: examples/mnist_net2net.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+        teacher_b1: `bias` of conv2d layer to become wider,
+          of shape (nb_filter1, )
+        teacher_w2: `weight` of next connected conv2d layer,
+          of shape (nb_filter2, nb_channel2, kh2, kw2)
+        new_width: new `nb_filter` for the wider conv2d layer
+        init: initialization algorithm for new weights,
+          either 'random-pad' or 'net2wider'
+    """
+    assert teacher_w1.shape[0] == teacher_w2.shape[1], (
+        "successive layers from teacher model should have compatible shapes")
```
Comment (fchollet) [inline examples/mnist_net2net.py]: You're using both `"` and `'` as the quote character, normalize to `'` everywhere.

File: examples/mnist_net2net.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+    student_b1 = np.concatenate((teacher_b1, new_b1), axis=0)
+
+    return student_w1, student_b1, student_w2
+
+
+def wider2net_fc(teacher_w1, teacher_b1, teacher_w2, new_width, init):
+    """Get initial weights for a wider fully connected (dense) layer
+       with a bigger nout, by 'random-padding' or 'net2wider'.
+
+    # Auguments
```
Comment (fchollet) [inline examples/mnist_net2net.py]: Typo

File: examples/mnist_net2net.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+    model.add(Conv2D(64, 3, 3, input_shape=input_shape,
+                     border_mode="same", name="conv1"))
+    model.add(MaxPooling2D(name="pool1"))
+    model.add(Conv2D(128, 3, 3, border_mode="same", name="conv2"))
+    model.add(MaxPooling2D(name="pool2"))
+    model.add(Flatten(name="flatten"))
+    model.add(Dense(128, activation="relu", name="fc1"))
+    model.add(Dense(nb_class, activation="softmax", name="fc2"))
+    model.compile(loss="categorical_crossentropy",
+                  optimizer="sgd", metrics=["accuracy"])
```
Comment (fchollet) [inline examples/mnist_net2net.py]: You shouldn't use SGD with default parameters. Prefer RMSprop or SGD with momentum and a tuned LR.

File: examples/mnist_net2net.py [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+    (3) a wider student model with `Net2WiderNet` initializer
+    """
+    train_data = (train_x, train_y)
+    validation_data = (validation_x, validation_y)
+    print("\nExperiment of Net2WiderNet ...")
+    print("\nbuilding teacher model ...")
+    teacher_model, _ = make_teacher_model(train_data, validation_data)
+
+    print("\nbuilding wider student model by random padding ...")
+    _ = make_wider_student_model(teacher_model, train_data,
```
Comment (fchollet) [inline examples/mnist_net2net.py]: `_ =` not necessary

File: examples/mnist_net2net.py (hunk 1 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 3 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 2 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 4 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 6 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 8 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: All changed to '''

File: examples/mnist_net2net.py (hunk 7 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

File: examples/mnist_net2net.py (hunk 9 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: all changed to '

File: examples/mnist_net2net.py (hunk 10 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed. sorry about it...

File: examples/mnist_net2net.py (hunk 11 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed to SGD with momentum=0.9, and different lr for different models, as suggested in the paper

File: examples/mnist_net2net.py (hunk 12 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: changed

**Comment by fchollet** [comment]
You don't seem to have pushed any changes.

File: examples/mnist_net2net.py (hunk 1 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: Change done. But the results running on GPU are still random.

**Comment by dolaameng** [comment]
Thanks for taking time making these comments -  good exercise for me to get familiar with the styles.

Main changes
1. Removing unnecessary white spaces
2. Fixing numpy.random.seed before Keras imports - but still got random results on GPU
3. Stick with single quotes across the example, including docstring
4. Typos fixed
5. Using SGD with momentum=0.9, and different lr for different models as discussed in the paper
6. Reporting validation loss, as training loss depends on the order and order of batches. 
7. Change of explanations in doc string to reflect the updated code.

File: examples/mnist_net2net.py [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+    train_x, train_y = train_data
+    history = model.fit(train_x, train_y, nb_epoch=nb_epoch,
+                        validation_data=validation_data)
+    return model, history
+
+
+# experiments setup
+def net2wider_experiment():
+    '''Benchmark performances of
+    (1) a teach model,
```
Comment (fchollet) [inline examples/mnist_net2net.py]: typo (teacher)

File: examples/mnist_net2net.py [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+    return student_w, student_b
+
+
+def copy_weights(teacher_model, student_model, layer_names):
+    '''Copy weights from teacher_model to student_model,
+     for layers with names listed in layer_names
+    '''
+    for name in layer_names:
+        weights = teacher_model.get_layer(name=name).get_weights()
+        student_model.get_layer(name=name).set_weights(weights)
```
Comment (fchollet) [inline examples/mnist_net2net.py]: This will be rather inefficient in TF, but overall it's okay.

**Comment by fchollet** [comment]
Looks good to me otherwise.

File: examples/mnist_net2net.py (hunk 14 above)
Comment (dolaameng) [inline examples/mnist_net2net.py]: Yeah. I will think about it. Also interested in exploring how to integrate this more naturally with Keras instead of getting out and in the weights...

**Comment by dolaameng** [comment]
Sorry hold it for this one. I will merge my fork with upstream first.

**Comment by dolaameng** [comment]
Sorry about creating build request. Will be more careful next time.

**Comment by fchollet** [comment]
LGTM
